File: CSharp/Grid.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MarsRoverKata
{
    class Grid
    {         
        //properties of grid
        public int maxX { get; private set; }
        public int minX { get; private set; }
        public int maxY { get; private set; }
        public int minY { get; private set; }

        public HashSet<Point> obstacles { get; private set; }  //collection of obstacle points

        //construct grid with given parameters
        public Grid(int maxX = 10, int minX = -10, int maxY = 10, int minY = -10)
        {
            this.maxX = maxX;
            this.minX = minX;
            this.maxY = maxY;
            this.minY = minY;

            //generate obstacles in 1/10th of the map
            int numObstacles = ((maxX - minX) * (maxY - minY)) / 10;
            obstacles = new HashSet<Point>();

            Random rand = new Random(); //to generate random numbers
            int index = 1;
            while(index <= numObstacles)
            {
                //generate random point; if it is a new point, add it to set
                Point p = new Point(rand.Next(minX, maxX + 1), rand.Next(minY, maxY + 1));
                if(!obstacles.Contains(p))
                {
                    obstacles.Add(p);
                    index++;
                }
            }
        }

        //method to show obstacle points
        public void showObstacles()
        {
            foreach (Point p in obstacles)
                Console.WriteLine("(" + p.X + ", " + p.Y + ")");
        }
    }
}

File: Ruby/grid.rb
require 'set'
require_relative 'point.rb'

class Grid
  def initialize (maxX = 10, minX = -10, maxY = 10, minY = -10)
    @maxX, @minX, @maxY, @minY = maxX, minX, maxY, minY

    @obstacles = Set.new
    numObstacles = ((maxX - minX) * (maxY - minY)) / 10

    (1..numObstacles).each do |n|
      p = Point.new(rand(minX..maxX), rand(minY..maxY))
      if !@obstacles.include?(p)
        @obstacles.add(p)
      end
    end
  end

  attr_reader :maxX, :minX, :maxY, :minY, :obstacles

  #method to show obstacles
  def showObstacles
    @obstacles.each do |p|
      puts "(#{p.x}, #{p.y})"
    end
  end
end